File: src/org/oopsc/statement/ReadStatement.java
package org.oopsc.statement;

import java.util.Stack;

import org.oopsc.*;
import org.oopsc.expression.*;
import org.oopsc.symbol.*;

/**
 * Die Klasse repräsentiert die Anweisung READ im Syntaxbaum.
 */
public class ReadStatement extends Statement {
	/** Die Variable, in der das eingelesene Zeichen gespeichert wird. */
	public Expression operand;

	/** Ein Ausdruck, der ein neues Objekt vom Typ Integer erzeugen kann. */
	public Expression newInt = new NewExpression(new ResolvableSymbol(
			new Identifier("Integer", new Position(0, 0)), null));

	/**
	 * Konstruktor.
	 *
	 * @param operand
	 *        Die Variable, in der das eingelesene Zeichen gespeichert wird.
	 */
	public ReadStatement(Expression operand) {
		this.operand = operand;
	}

	@Override
	public void refPass(SemanticAnalysis sem) throws CompileException {
		this.operand = this.operand.refPass(sem);

		if (!this.operand.lValue) {
			throw new CompileException("Lvalue expected", this.operand.position);
		}

		this.operand.type.check(sem, sem.types().intClass(),
				this.operand.position);
		this.newInt = this.newInt.refPass(sem);
	}

	@Override
	public void print(TreeStream tree) {
		tree.println("READ");
		tree.indent();
		this.operand.print(tree);
		tree.unindent();
	}

	@Override
	public void generateCode(CodeStream code, Stack<Context> contexts) {
		// TODO refactor
		code.println("; READ lvalue ablegen");
		this.operand.generateCode(code);

		code.println("; READ Speicher allokieren");
		this.newInt.generateCode(code);

		code.println("; READ");
		code.println("MRM R5, (R2)"); // R2 zeigt auf ein boxed Integer

		/* Skip header. */
		code.println("MRI R6, " + ClassSymbol.HEADERSIZE());
		code.println("ADD R5, R6");

		code.println("SYS 0, 6 ; Gelesenen Wert in R6 ablegen");
		code.println("MMR (R5), R6 ; Zeichen in neuen Integer schreiben");
		code.println("MRM R5, (R2) ; Neuen Integer vom Stapel entnehmen");
		code.println("SUB R2, R1");
		code.println("MRM R6, (R2) ; Ziel vom Stapel entnehmen");
		code.println("SUB R2, R1");
		code.println("MMR (R6), R5 ; Zuweisen");
	}
}

File: src/org/oopsc/statement/ReadStatement.scala
package org.oopsc.statement

import org.oopsc._
import org.oopsc.expression._
import org.oopsc.symbol._

/**
 * Reads a character and stores it in the given operand.
 */
class ReadStatement(var operand: Expression) extends Statement {
  /** An expression for creating a new object with the type Integer. */
  private var newInt = new NewExpression(new ResolvableClassSymbol(new Identifier("Integer")))

  override def refPass(sem: SemanticAnalysis) {
    this.operand.refPass(sem)

    if (!this.operand.lValue) {
      throw new CompileException("Variable reference expected.", this.operand.position)
    }

    this.operand.resolvedType.check(Types.intClass, this.operand.position)
    this.newInt.refPass(sem)
  }

  override def optimPass() : Statement = {
    this.operand = this.operand.optimPass()
    this
  }

  override def print(tree: TreeStream) {
    tree.println("READ")
    tree.indent
    this.operand.print(tree)
    tree.unindent
  }

  override def generateCode(code: CodeStream, tryContexts: Int) {
    code.println("; READ")
    code.println("; Push operand (lvalue) on the stack.")
    this.operand.generateCode(code)

    code.println("; Allocate memory for the character.")
    this.newInt.generateCode(code)

    code.println("; READ")
    code.println("MRM R5, (R2)") /* R2 points to a boxed Integer value. */

    /* Skip header. */
    code.println(s"MRI R6, ${ClassSymbol.HEADERSIZE}")
    code.println("ADD R5, R6")

    code.println("SYS 0, 6 ; Store read value in R6.")
    code.println("MMR (R5), R6 ; Set the value of the Integer object to the read character.")
    code.println("MRM R5, (R2) ; Read the allocated object reference from the stack.")
    code.println("SUB R2, R1")
    code.println("MRM R6, (R2) ; Read destination from the stack.")
    code.println("SUB R2, R1")
    code.println("MMR (R6), R5 ; Assign.")
  }
}